src/core/ddsi/src/ddsi_init.c

```diff
@@ -1207,48 +1244,104 @@ int ddsi_init (struct ddsi_domaingv *gv, struct ddsi_psmx_instance_locators *psm
+  // UnicastSPDPInterval -> 30s       |_ perhaps adding something like this
+  //   @silentports -> 5min           |  would make sense?
+  //   @dropafter -> 30min           -+
+  bool add_self_to_as_disc = false;
+  if (gv->config.participantIndex == DDSI_PARTICIPANT_INDEX_DEFAULT)
   {
+#ifndef NDEBUG
     for (int i = 0; i < gv->n_interfaces; i++)
     {
-      if (!gv->interfaces[i].mc_capable)
-      {
-        GVWARNING ("selected interface \"%s\" is not multicast-capable: disabling multicast\n", gv->interfaces[i].name);
-        gv->config.allowMulticast = DDSI_AMC_FALSE;
-        /* ensure discovery can work: firstly, that the process will be reachable on a "well-known" port
-         number, and secondly, that the local interface's IP address gets added to the discovery
-         address set */
-        gv->config.participantIndex = DDSI_PARTICIPANT_INDEX_AUTO;
-      }
-      else if (gv->config.allowMulticast & DDSI_AMC_DEFAULT)
-      {
-        /* default is dependent on network interface type: if multicast is believed to be flaky,
-         use multicast only for SPDP packets */
-        assert ((gv->config.allowMulticast & ~DDSI_AMC_DEFAULT) == 0);
-        if (gv->interfaces[i].mc_flaky)
-        {
-          gv->config.allowMulticast = DDSI_AMC_SPDP;
-          GVLOG (DDS_LC_CONFIG, "presumed flaky multicast, use for SPDP only\n");
-        }
-        else
-        {
-          GVLOG (DDS_LC_CONFIG, "presumed robust multicast support, use for everything\n");
-          gv->config.allowMulticast = DDSI_AMC_TRUE;
-        }
-      }
+      // sanity check that by now we have eliminated "default" from allow_multicast and
+      // that no bits in allow_multicast are set if the interface is not capable of
+      // handling multicast
+      assert ((gv->interfaces[i].allow_multicast & DDSI_AMC_DEFAULT) == 0);
+      assert (gv->interfaces[i].allow_multicast == 0 || gv->interfaces[i].mc_capable);
+    }
+#endif
+    bool all_allow_spdp_mc = true, none_allow_spdp_mc = true;
+    for (int i = 0; i < gv->n_interfaces; i++)
+    {
+      if (gv->interfaces[i].allow_multicast & DDSI_AMC_SPDP)
+        none_allow_spdp_mc = false;
+      else
+        all_allow_spdp_mc = false;
+    }
+    if (all_allow_spdp_mc && gv->config.peers == NULL)
+    {
+      GVTRACE ("all interfaces allow spdp multicast, no peers defined: defaulting participant index to \"none\"\n");
+      gv->config.participantIndex = DDSI_PARTICIPANT_INDEX_NONE;
+    } else if (all_allow_spdp_mc)
+    {
+      GVTRACE ("all interfaces allow spdp multicast, but peers defined: defaulting participant index to \"auto\"\n");
+      gv->config.participantIndex = DDSI_PARTICIPANT_INDEX_AUTO;
+    }
+    else
+    {
+      GVTRACE ("some interfaces disallow spdp multicast: defaulting participant index to \"auto\"\n");
+      gv->config.participantIndex = DDSI_PARTICIPANT_INDEX_AUTO;
+    }
+    if (none_allow_spdp_mc && !gv->config.config_has_empty_peers_section)
```

Comment:
I thought about it a bit and decided that it would be better to have less implicit behaviour: 65a8843 adds an attribute to the "Peers" element, which I believe makes it both more flexible and clearer.

Reply:
Yes, this way it's clear what the setting does and what's the reason for adding localhost as peer.